### PR TITLE
Add: support dns (lookup) command

### LINKED ISSUE #2: DNS command - Real time update
DNS command measures resolving performance and collects resolved output.
See https://github.com/jsdelivr/globalping/issues/6 for command options

A possible command that can be used to resolve target:

```
dig -4 +time=1 +tries=2 +edns +stats @1.1.1.1 MX google.com
```

dig output is pretty complex and may change significantly depending on the params so we should avoid parsing it manually.

Need to check this package https://www.npmjs.com/package/node-dig-dns and see if it can parse the output with the params we need.

### PR DESCRIPTION
**Pull request by patrykcieszkowski**
resolve #2

There's not many actively maintained libraries around, since node's official package resolves DNS pretty well (https://nodejs.org/api/dns.html).

It doesn't however seem to support protocol selection, so that won't work for us.

My initial commit implemented [domain-info](https://www.npmjs.com/package/domain-info) package, but I decided to ditch it since it doesn't seem to be actively maintained either, its last build failed. It supports multi dns type queries, which is interesting. The package @zarianec ([node-dig-dns](https://www.npmjs.com/package/node-dig-dns)) has proposed on the other hand, doesn't include types so they had to be included manually.

No tests, since theres nothing to test. Parsing is handled by the library.

### CODE REVIEW

**Comment by patrykcieszkowski** [comment]
@zarianec I don't think this DeepSource linter is configured in accordance with our project eslint.

**Comment by jimaek** [comment]
> @zarianec I don't think this DeepSource linter is configured in accordance with our project eslint.

You're admin of deepsource. Can you please configure it as you see fit?

**Comment by zarianec** [comment]
yep, we just enabled it so it may require rules tweaking to make it work as expected. It's not necessary DeepSource checks to always pass. The main goal is to provide additional information and insights.
If you see that some errors could be ignored or they are false positives - ignore them. Or even better - disable them in the dashboard. But be careful there and try to not disable some correct rules.

**Comment by zarianec** [comment]
and some rules in DeepSource are really stupid I must say 🤯

File: src/command/dns-command.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+};
+
+const allowedTypes = ['A', 'AAAA', 'ANY', 'CNAME', 'DNSKEY', 'DS', 'MX', 'NS', 'NSEC', 'PTR', 'RRSIG', 'SOA', 'TXT', 'SRV'];
+
+const dnsOptionsSchema = Joi.object<DnsOptions>({
+	type: Joi.string().valid('dns'),
+	target: Joi.string(),
+	query: Joi.object({
+		type: Joi.string().valid(...allowedTypes).optional(),
+		address: Joi.string().optional(),
```
Comment (zarianec) [inline src/command/dns-command.ts:26]: resolver or nameServer should be a better name.

File: src/command/dns-command.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		port: Joi.number().optional(),
+	}),
+});
+
+export const dnsCmd = async (options: DnsOptions): Promise<DnsQueryResult> => {
+	const protocolArg = options.query.protocol?.toLowerCase() === 'tcp' ? '+tcp' : '';
+
+	const args = [
+		options.target,
+		`@${options.query.address ?? getDnsServers().pop()!}`,
```
Comment (zarianec) [inline src/command/dns-command.ts:37]: This is an optimistic approach 🙂 . Dns servers returned by `dns.getServers()` may include addresses with a port as well as ipv6 addresses. Why not just avoid the `@servername` option if no `options.query.address` provided?

File: src/command/dns-command.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		['-t', options.query.type ?? 'A'],
+		['-p', options.query.port ?? '53'],
```
Comment (zarianec) [inline src/command/dns-command.ts:39]: `-4 +time=1 +tries=2` params missing. They must be hardcoded.

**Review by zarianec — CHANGES_REQUESTED** [review]

**Comment by zarianec** [comment]
I guess it's not possible with `node-dig-dns` to get the command progress? We may want to add `+trace` later for tracing the resolution and it will require real-time updates.

File: src/command/dns-command.ts (hunk 2 above)
Comment (patrykcieszkowski) [reply src/command/dns-command.ts:37]: Maybe. That's a leftover from the previous library - it wouldn't find its own defaults.

**Comment by patrykcieszkowski** [comment]
> I guess it's not possible with `node-dig-dns` to get the command progress? We may want to add `+trace` later for tracing the resolution and it will require real-time updates.

It doesn't, it outputs a promise resolution after completed query. If we want the progress event to be outputted, we'll have to build our own library (or just copy the codebase of `node-dig-dns` and adapt for our needs)

File: src/command/dns-command.ts (hunk 2 above)
Comment (jimaek) [reply src/command/dns-command.ts:37]: We should let dig use the system defaults for now, so dont provide anything manually

**Comment by zarianec** [comment]
@patrykcieszkowski tests missing for this new command.

**Comment by zarianec** [comment]
@patrykcieszkowski tests missing for this new command.

**Comment by patrykcieszkowski** [comment]
> @patrykcieszkowski tests missing for this new command.

read the top message

**Comment by zarianec** [comment]
> No tests, since theres nothing to test. Parsing is handled by the library.

Not true. Yes, parsing is done by a 3rd-party library but we send this output to the API using websockets and we must be sure that this output won't change after we update the library, change the code, etc.
What's important in commands - is the messages they send to the API. Parsing is just a part of it.

**Comment by zarianec** [comment]
My goal is eventually to have 100% code coverage.

**Comment by zarianec** [comment]
Following the discussion in private channels, it's not much to test there since we use an external library and don't have a mechanism of mocking/testing calls for 3rd-party system components (ping, dig, etc.).
I'm going to merge this PR and we will figure out how to properly test it in a separate one.

**Pull request merged**